File: src/main/java/com/sotasan/decompiler/transformers/Transformer.java
package com.sotasan.decompiler.transformers;

import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;

public enum Transformer {

    CFR(CFRTransformer.class),
    Procyon(ProcyonTransformer.class),
    Vineflower(VineflowerTransformer.class);

    private final Class<? extends ITransformer> clazz;

    Transformer(Class<? extends ITransformer> clazz) {
        this.clazz = clazz;
    }

    @SneakyThrows
    public @NotNull ITransformer newInstance() {
        return clazz.getConstructor().newInstance();
    }

}

File: src/main/kotlin/moe/sota/decompiler/transformers/Transformer.kt
package moe.sota.decompiler.transformers

enum class Transformer(private val factory: () -> ITransformer) {
    CFR(::CFRTransformer),
    JD(::JDTransformer),
    Procyon(::ProcyonTransformer),
    Vineflower(::VineflowerTransformer);

    fun newInstance(): ITransformer = factory()
}
